safe_transaction_service/history/indexers/safe_events_indexer.py

```diff
@@ -485,3 +478,137 @@ def _process_decoded_element(
+        self, decoded_elements: List[EventData]
+    ) -> Dict[ChecksumAddress, List[EventData]]:
+        """
+        Get the creation events (ProxyCreation and SafeSetup) from decoded elements and generates a dictionary that groups these events by Safe address.
+
+        :param decoded_elements:
+        :return:
+        """
+        safe_setup_events: Dict[ChecksumAddress, List[Dict]] = {}
+        for decoded_element in decoded_elements:
+            event_name = decoded_element["event"]
+            if event_name == "SafeSetup":
+                safe_address = decoded_element["address"]
+                safe_setup_events.setdefault(safe_address, []).append(decoded_element)
+            elif event_name == "ProxyCreation":
+                safe_address = decoded_element["args"].get("proxy")
+                safe_setup_events.setdefault(safe_address, []).append(decoded_element)
+
+        return safe_setup_events
+
+    @transaction.atomic
+    def _process_safe_creation_events(
+        self, safe_setup_events: Dict[ChecksumAddress, List[EventData]]
+    ) -> List[InternalTx]:
+        """
+        Process creation events (ProxyCreation and SafeSetup).
+
+        :param safe_setup_events:
+        :return:
+        """
+        internal_txs = []
+        internal_decoded_txs = []
+        # Check if were indexed
+        safe_setup_events_addresses = list(safe_setup_events.keys())
+        indexed_addresses = InternalTxDecoded.objects.filter(
+            internal_tx___from__in=safe_setup_events_addresses,
+            function_name="setup",
+            internal_tx__contract_address=None,
+        ).values_list("internal_tx___from", flat=True)
+        addresses_to_index = set(safe_setup_events_addresses) - set(indexed_addresses)
+        for safe_address in addresses_to_index:
+            events = safe_setup_events[safe_address]
+            for event_position, event in enumerate(events):
+                if event["event"] == "SafeSetup":
+                    setup_event = event
+                    # If we have both events we should extract Singleton and trace_address from ProxyCreation event
+                    if len(events) > 1:
+                        if (
+                            event_position == 0
+                            and events[1]["event"] == "ProxyCreation"
+                        ):
+                            # Usually SafeSetup is the first event and next is ProxyCreation when ProxyCreation is called with initializer.
+                            proxy_creation_event = events[1]
+                        elif (
+                            event_position == 1
+                            and events[0]["event"] == "ProxyCreation"
+                        ):
+                            # ProxyCreation first and SafeSetup later
+                            proxy_creation_event = events[1]
+                    else:
+                        # Proxy was created in previous blocks.
```

Comment:
The previous implementation always tried to remove the `InternalTx` generated from `ProxyCreation` just in case the `ProxyCreation` came first. https://github.com/safe-global/safe-transaction-service/blob/main/safe_transaction_service/history/indexers/safe_events_indexer.py#L366 
 In this case, because the `ProxyCreation` never comes later, the `InternalTx` generated by `SafeSetup` leaves without singleton (to) forever.  
 Maybe would be better try to get the Proxy InternalTx from database to insert the correct parameters of `SafeSetup`  what do you think? @safe-global/core-api 
 In the current implementation the `ProxyCreation` is storing the singleton addres in the `to` field of `InternalTx`, not sure if is completely right.